File: app/src/main/java/com/vasu/assistant/core/voice/GeminiLiveSession.kt
package com.vasu.assistant.core.voice

import android.util.Base64
import android.util.Log
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject
import javax.inject.Singleton

/**
 * GeminiLiveSession - Bidirectional real-time WebSocket session coordinator
 * for Google Gemini Live Multimodal Audio API.
 *
 * Voice: "Kore"
 * Audio In: 16-bit Mono PCM @ 16 kHz
 * Audio Out: 16-bit Mono PCM @ 24 kHz
 */
@Singleton
class GeminiLiveSession @Inject constructor() {

    companion object {
        private const val TAG = "GeminiLiveSession"
        private const val WS_HOST = "wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1alpha.GenerativeService.BidiGenerateContent"
        const val TARGET_VOICE = "Kore"
        const val LIVE_MODEL = "models/gemini-2.0-flash-exp"
    }

    private var onAudioReceived: ((ByteArray) -> Unit)? = null
    private var onTextReceived: ((String) -> Unit)? = null
    private var onInterrupted: (() -> Unit)? = null
    private var onTurnComplete: (() -> Unit)? = null
    private var onError: ((String) -> Unit)? = null

    private val isConnected = AtomicBoolean(false)
    private val isReadyState = AtomicBoolean(false)

    private var webSocket: WebSocket? = null
    private var okHttpClient: OkHttpClient? = null
    private var readyDeferred: CompletableDeferred<Boolean>? = null

    fun setCallbacks(
        onAudioReceived: (ByteArray) -> Unit,
        onTextReceived: (String) -> Unit,
        onInterrupted: () -> Unit,
        onTurnComplete: () -> Unit,
        onError: (String) -> Unit
    ) {
        this.onAudioReceived = onAudioReceived
        this.onTextReceived = onTextReceived
        this.onInterrupted = onInterrupted
        this.onTurnComplete = onTurnComplete
        this.onError = onError
    }

    fun isReady(): Boolean = isConnected.get() && isReadyState.get()

    /**
     * Connect to Gemini Live WebSocket and suspend until setup handshake completes.
     */
    suspend fun connectAndWaitReady(apiKey: String, systemInstruction: String): Boolean = withContext(Dispatchers.IO) {
        disconnect()

        val deferred = CompletableDeferred<Boolean>()
        readyDeferred = deferred

        try {
            val client = OkHttpClient.Builder()
                .readTimeout(0, TimeUnit.MILLISECONDS)
                .pingInterval(20, TimeUnit.SECONDS)
                .retryOnConnectionFailure(true)
                .build()
            okHttpClient = client

            val url = "$WS_HOST?key=${apiKey.trim()}"
            val request = Request.Builder().url(url).build()

            webSocket = client.newWebSocket(request, object : WebSocketListener() {
                override fun onOpen(ws: WebSocket, response: Response) {
                    Log.d(TAG, "Gemini Live WebSocket opened; sending setup payload")
                    isConnected.set(true)
                    val setupMessage = buildSetupMessage(systemInstruction)
                    ws.send(setupMessage)
                }

                override fun onMessage(ws: WebSocket, text: String) {
                    handleIncomingServerMessage(text)
                }

                override fun onClosing(ws: WebSocket, code: Int, reason: String) {
                    Log.d(TAG, "Gemini Live closing: $code / $reason")
                    isConnected.set(false)
                    isReadyState.set(false)
                    ws.close(1000, null)
                }

                override fun onClosed(ws: WebSocket, code: Int, reason: String) {
                    Log.d(TAG, "Gemini Live closed: $code / $reason")
                    isConnected.set(false)
                    isReadyState.set(false)
                    readyDeferred?.complete(false)
                }

                override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
                    val err = t.message ?: "WebSocket failure"
                    Log.e(TAG, "Gemini Live failure: $err", t)
                    isConnected.set(false)
                    isReadyState.set(false)
                    readyDeferred?.complete(false)
                    onError?.invoke(err)
                }
            })

            val ready = withTimeoutOrNull(8000) {
                deferred.await()
            } ?: false

            if (!ready) {
                Log.w(TAG, "Timed out waiting for Gemini Live setup confirmation")
            }
            return@withContext ready
        } catch (e: Exception) {
            Log.e(TAG, "Exception connecting to Gemini Live", e)
            readyDeferred?.complete(false)
            onError?.invoke(e.message ?: "Failed to connect to Gemini Live")
            return@withContext false
        }
    }

    private fun buildSetupMessage(systemInstruction: String): String {
        val root = JSONObject()
        val setup = JSONObject()
        setup.put("model", LIVE_MODEL)

        val generationConfig = JSONObject()
        val responseModalities = JSONArray()
        responseModalities.put("AUDIO")
        generationConfig.put("responseModalities", responseModalities)

        val speechConfig = JSONObject()
        val voiceConfig = JSONObject()
        val prebuiltVoiceConfig = JSONObject()
        prebuiltVoiceConfig.put("voiceName", TARGET_VOICE) // "Kore"
        voiceConfig.put("prebuiltVoiceConfig", prebuiltVoiceConfig)
        speechConfig.put("voiceConfig", voiceConfig)
        generationConfig.put("speechConfig", speechConfig)

        setup.put("generationConfig", generationConfig)

        if (systemInstruction.isNotBlank()) {
            val systemInstructionObj = JSONObject()
            val parts = JSONArray()
            val part = JSONObject()
            part.put("text", systemInstruction)
            parts.put(part)
            systemInstructionObj.put("parts", parts)
            setup.put("systemInstruction", systemInstructionObj)
        }

        root.put("setup", setup)
        return root.toString()
    }

    private fun handleIncomingServerMessage(jsonText: String) {
        try {
            val root = JSONObject(jsonText)

            // Setup complete handshake
            if (root.has("setupComplete")) {
                Log.d(TAG, "Gemini Live session setupComplete received - session is READY")
                isReadyState.set(true)
                readyDeferred?.complete(true)
                return
            }

            val serverContent = root.optJSONObject("serverContent")
            if (serverContent != null) {
                if (serverContent.optBoolean("interrupted", false)) {
                    Log.d(TAG, "Server reported user interrupted assistant")
                    onInterrupted?.invoke()
                }

                val modelTurn = serverContent.optJSONObject("modelTurn")
                if (modelTurn != null) {
                    val parts = modelTurn.optJSONArray("parts")
                    if (parts != null) {
                        for (i in 0 until parts.length()) {
                            val part = parts.getJSONObject(i)

                            // Native AUDIO chunk
                            val inlineData = part.optJSONObject("inlineData")
                            if (inlineData != null) {
                                val base64Audio = inlineData.optString("data", "")
                                if (base64Audio.isNotEmpty()) {
                                    val pcmBytes = Base64.decode(base64Audio, Base64.DEFAULT)
                                    onAudioReceived?.invoke(pcmBytes)
                                }
                            }

                            // Model text transcript
                            if (part.has("text")) {
                                val text = part.optString("text", "")
                                if (text.isNotEmpty()) {
                                    onTextReceived?.invoke(text)
                                }
                            }
                        }
                    }
                }

                if (serverContent.optBoolean("turnComplete", false)) {
                    Log.d(TAG, "Model turn complete")
                    onTurnComplete?.invoke()
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error handling Gemini Live server payload", e)
        }
    }

    /**
     * Send user text prompt to Gemini Live to synthesize real native Kore audio response.
     */
    fun sendTextTurn(promptText: String): Boolean {
        if (!isReady()) {
            Log.w(TAG, "Cannot send text: session not ready")
            return false
        }
        return try {
            val root = JSONObject()
            val clientContent = JSONObject()
            val turns = JSONArray()
            val turn = JSONObject()
            turn.put("role", "user")
            val parts = JSONArray()
            val part = JSONObject()
            part.put("text", promptText)
            parts.put(part)
            turn.put("parts", parts)
            turns.put(turn)
            clientContent.put("turns", turns)
            clientContent.put("turnComplete", true)
            root.put("clientContent", clientContent)

            webSocket?.send(root.toString()) ?: false
        } catch (e: Exception) {
            Log.e(TAG, "Error sending text turn to Gemini Live", e)
            false
        }
    }

    /**
     * Stream user microphone audio (16-bit PCM @ 16 kHz mono) into Gemini Live session.
     */
    fun sendAudioChunk(pcmChunk: ByteArray): Boolean {
        if (!isReady() || pcmChunk.isEmpty()) return false
        return try {
            val root = JSONObject()
            val realtimeInput = JSONObject()
            val mediaChunks = JSONArray()
            val chunk = JSONObject()
            chunk.put("mimeType", "audio/pcm;rate=16000")
            chunk.put("data", Base64.encodeToString(pcmChunk, Base64.NO_WRAP))
            mediaChunks.put(chunk)
            realtimeInput.put("mediaChunks", mediaChunks)
            root.put("realtimeInput", realtimeInput)

            webSocket?.send(root.toString()) ?: false
        } catch (e: Exception) {
            Log.e(TAG, "Error sending realtime audio chunk", e)
            false
        }
    }

    fun disconnect() {
        isConnected.set(false)
        isReadyState.set(false)
        readyDeferred?.complete(false)
        try {
            webSocket?.close(1000, "Normal closure")
            webSocket = null
        } catch (e: Exception) {
            Log.w(TAG, "Error closing websocket", e)
        }
    }
}

File: src/services/geminiLiveSession.ts
/**
 * GeminiLiveSession - Manages WebSocket connection to Gemini Live API.
 *
 * Lifecycle:
 * CONNECTING -> OPEN -> SESSION CONFIGURED -> READY -> STREAMING
 *
 * Enforces:
 * - Never calls websocket.send() before the socket is OPEN.
 * - Every send operation verifies the real OPEN state.
 * - Handles onopen, onmessage, onerror, onclose cleanly.
 * - No unhandled Promise rejection is allowed.
 */

export type LiveSessionState =
  | 'DISCONNECTED'
  | 'CONNECTING'
  | 'OPEN'
  | 'SESSION_CONFIGURED'
  | 'READY'
  | 'STREAMING'
  | 'ERROR';

export interface GeminiLiveSessionCallbacks {
  onStateChange: (state: LiveSessionState) => void;
  onAudioChunk: (pcmBytes: Uint8Array) => void;
  onTranscript: (text: string) => void;
  onInterrupted: () => void;
  onTurnComplete?: () => void;
  onError: (errorMsg: string) => void;
}

export class GeminiLiveSession {
  private ws: WebSocket | null = null;
  private state: LiveSessionState = 'DISCONNECTED';
  private callbacks: GeminiLiveSessionCallbacks;

  public static readonly TARGET_VOICE = 'Erinome';
  public static readonly LIVE_MODEL = 'models/gemini-2.0-flash-exp';

  private readyResolver: ((success: boolean) => void) | null = null;

  constructor(callbacks: GeminiLiveSessionCallbacks) {
    this.callbacks = callbacks;
  }

  public getState(): LiveSessionState {
    return this.state;
  }

  private setState(newState: LiveSessionState): void {
    this.state = newState;
    this.callbacks.onStateChange(newState);
  }

  /**
   * Connect to Gemini Live WebSocket.
   * Can connect directly via API key, or via the local server proxy (/api/live-ws).
   * Resolves only after WebSocket is OPEN, setup is sent, and setupComplete is received (READY).
   */
  public connect(apiKey?: string, systemPrompt?: string): Promise<boolean> {
    return new Promise<boolean>((resolve) => {
      try {
        if (this.ws && this.ws.readyState === WebSocket.OPEN && this.state === 'READY') {
          console.log('[VASU] Gemini session already OPEN and READY');
          resolve(true);
          return;
        }

        this.disconnect();

        this.readyResolver = resolve;
        this.setState('CONNECTING');
        console.log('[VASU] Gemini session connecting');

        // Determine URL: direct or proxy
        let url: string;
        if (apiKey && apiKey.trim()) {
          url = `wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1alpha.GenerativeService.BidiGenerateContent?key=${encodeURIComponent(apiKey.trim())}`;
        } else {
          // Connect to backend server proxy
          const loc = window.location;
          const proto = loc.protocol === 'https:' ? 'wss:' : 'ws:';
          url = `${proto}//${loc.host}/api/live-ws`;
        }

        const socket = new WebSocket(url);
        this.ws = socket;

        socket.onopen = () => {
          this.setState('OPEN');
          console.log('[VASU] Gemini session OPEN');

          // Send setup configuration immediately once OPEN
          this.sendSetup(systemPrompt);
        };

        socket.onmessage = (event) => {
          this.handleMessage(event.data);
        };

        socket.onerror = (event) => {
          console.error('[VASU] Gemini session WebSocket error:', event);
          this.setState('ERROR');
          this.callbacks.onError('WebSocket connection error');
          if (this.readyResolver) {
            this.readyResolver(false);
            this.readyResolver = null;
          }
        };

        socket.onclose = (event) => {
          console.log(`[VASU] Gemini session closed (code=${event.code}, wasClean=${event.wasClean})`);
          this.setState('DISCONNECTED');
          if (this.readyResolver) {
            this.readyResolver(false);
            this.readyResolver = null;
          }
        };
      } catch (err: any) {
        console.error('[VASU] Exception during WebSocket connection setup:', err?.message || err);
        this.setState('ERROR');
        this.callbacks.onError(err?.message || 'Connection failed');
        if (this.readyResolver) {
          this.readyResolver(false);
          this.readyResolver = null;
        }
      }
    });
  }

  /**
   * Send initial session configuration to Gemini Live.
   */
  private sendSetup(customSystemInstruction?: string): void {
    if (!this.isOpen()) {
      console.warn('[VASU] Cannot send setup: socket is not OPEN');
      return;
    }

    const defaultInstruction =
      'You are VASU, a natural real-time voice assistant.\n' +
      'Speak naturally and conversationally.\n' +
      "Respond in the user's language.\n" +
      'If the user speaks Hindi, respond in Hindi.\n' +
      'If the user speaks Hinglish, respond naturally in Hinglish.\n' +
      'Keep simple answers concise.\n' +
      "Do not unnecessarily repeat the user's words.\n" +
      'Do not mention internal APIs, models, WebSockets, or implementation details.\n\n' +
      'तुम VASU हो, एंड्रॉइड के लिए एक कुशल, स्नेही और बुद्धिमान AI वॉइस असिस्टेंट।\n' +
      'सामान्य बातचीत में हमेशा स्वाभाविक, बोलचाल की भाषा में उत्तर दो।\n' +
      'उत्तर संक्षिप्त, स्पष्ट एवं स्वाभाविक रखो जिसे सीधे बोला जा सके।';

    const systemText = (customSystemInstruction && customSystemInstruction.trim()) || defaultInstruction;

    const setupMessage = {
      setup: {
        model: GeminiLiveSession.LIVE_MODEL,
        generationConfig: {
          responseModalities: ['AUDIO'],
          speechConfig: {
            voiceConfig: {
              prebuiltVoiceConfig: {
                voiceName: GeminiLiveSession.TARGET_VOICE,
              },
            },
          },
        },
        systemInstruction: {
          parts: [{ text: systemText }],
        },
      },
    };

    const sent = this.sendRaw(JSON.stringify(setupMessage));
    if (sent) {
      this.setState('SESSION_CONFIGURED');
      console.log('[VASU] Gemini session configured');
    }
  }

  private handleMessage(data: string | ArrayBuffer): void {
    try {
      if (typeof data !== 'string') {
        return;
      }

      const message = JSON.parse(data);

      if (message._vasu_proxy === 'UPSTREAM_OPEN') {
        // Backend proxy upstream connection ready
        return;
      }

      // 1. Setup complete -> READY
      if (message.setupComplete) {
        this.setState('READY');
        console.log('[VASU] Gemini session setupComplete: READY');
        if (this.readyResolver) {
          this.readyResolver(true);
          this.readyResolver = null;
        }
        return;
      }

      // 2. Server content
      if (message.serverContent) {
        const { modelTurn, turnComplete, interrupted } = message.serverContent;

        if (interrupted) {
          console.log('[VASU] Gemini interruption received');
          this.callbacks.onInterrupted();
        }

        if (modelTurn && Array.isArray(modelTurn.parts)) {
          for (const part of modelTurn.parts) {
            // Text transcription
            if (part.text) {
              this.callbacks.onTranscript(part.text);
            }

            // Native 24 kHz 16-bit PCM Audio chunk
            if (part.inlineData && part.inlineData.data) {
              const rawData = part.inlineData.data;
              const binary = atob(rawData);
              const bytes = new Uint8Array(binary.length);
              for (let i = 0; i < binary.length; i++) {
                bytes[i] = binary.charCodeAt(i);
              }
              console.log('[VASU] Audio response received');
              console.log(`[VASU] Audio bytes received: ${bytes.byteLength}`);
              this.callbacks.onAudioChunk(bytes);
            }
          }
        }

        if (turnComplete) {
          this.callbacks.onTurnComplete?.();
        }
      }
    } catch (err: any) {
      console.error('[VASU] Failed to parse Gemini Live message:', err?.message || err);
    }
  }

  /**
   * Send text turn (e.g. for TEXT-ONLY test).
   */
  public sendText(text: string): boolean {
    if (!this.isOpen()) {
      console.warn('[VASU] Cannot send text: socket is not OPEN');
      return false;
    }

    const clientContent = {
      clientContent: {
        turns: [
          {
            role: 'user',
            parts: [{ text: text }],
          },
        ],
        turnComplete: true,
      },
    };

    return this.sendRaw(JSON.stringify(clientContent));
  }

  /**
   * Send real-time microphone 16 kHz 16-bit PCM audio chunk.
   */
  public sendAudioChunk(base64PcmChunk: string): boolean {
    if (!this.isOpen()) {
      return false;
    }

    const realtimeInput = {
      realtimeInput: {
        mediaChunks: [
          {
            mimeType: 'audio/pcm;rate=16000',
            data: base64PcmChunk,
          },
        ],
      },
    };

    const sent = this.sendRaw(JSON.stringify(realtimeInput));
    if (sent) {
      this.setState('STREAMING');
      console.log('[VASU] Audio chunk sent');
    }
    return sent;
  }

  /**
   * Low-level send with strict OPEN verification.
   */
  private sendRaw(data: string): boolean {
    if (!this.isOpen()) {
      console.warn('[VASU] sendRaw dropped message: WebSocket is not in OPEN state');
      return false;
    }
    try {
      this.ws!.send(data);
      return true;
    } catch (err: any) {
      console.error('[VASU] Error sending WebSocket data:', err?.message || err);
      return false;
    }
  }

  public isOpen(): boolean {
    return this.ws !== null && this.ws.readyState === WebSocket.OPEN;
  }

  public disconnect(): void {
    if (this.ws) {
      try {
        this.ws.close(1000, 'Normal disconnect');
      } catch {
        // Ignored
      }
      this.ws = null;
    }
    this.setState('DISCONNECTED');
  }
}
